In [294]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [295]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [296]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [297]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [298]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file

#Select numerical columns to scale (replace 'column1', 'column2', etc. with your actual column names)
numerical_cols = df_market_data.select_dtypes(include=['float64', 'int64']).columns
# Initialize the StandardScaler
scaler = StandardScaler()


In [299]:
# Create a DataFrame with the scaled data

# Fit the scaler to the data and transform it
scaled_data = scaler.fit_transform(df_market_data[numerical_cols])

# Create a DataFrame with the scaled data
df_scaled = pd.DataFrame(scaled_data, columns=numerical_cols, index=df_market_data.index)
# Display the first few rows of the scaled DataFrame
df_scaled.head()


# # Copy the crypto names from the original DataFrame (my code already does it)
# Set the coin_id column as index

# Display the scaled DataFrame
df_scaled.index
df_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [300]:
# Create a list with the number of k-values from 1 to 11
k_values = range(1, 11)  # You can adjust the range as needed


In [301]:
# Create an empty list to store the inertia values
# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list


# Create a list to store inertia values and the range of k values
inertia = []

# Loop through the range of k values
for k in k_values:
    # Create a KMeans model using the loop counter for the n_clusters
    model = KMeans(n_clusters=k, random_state=42)
    
    # 2. Fit the model to the data using `df_market_data_scaled`
    model.fit(df_scaled)
    
    # 3. Append the model.inertia_ to the inertia list
    inertia.append(model.inertia_)


C:\Users\danie\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\danie\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\danie\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\danie\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is k

In [302]:
# Create a dictionary with the data to plot the Elbow curve

#zip: pairs each ( k ) value with its corresponding inertia value, creating tuples of the form (k, inertia).
#dict: This converts the zipped tuples into a dictionary.
elbow_data_dict = dict(zip(k_values, inertia))

# Create a DataFrame with the data to plot the Elbow curve
elbow_data_df = pd.DataFrame(list(elbow_data_dict.items()), columns=['k', 'inertia'])
elbow_data_df



k     inertia
0   1  287.000000
1   2  212.123342
2   3  170.193571
3   4   79.022435
4   5   71.097914
5   6   66.108827
6   7   54.488123
7   8   43.224693
8   9   32.860884
9  10   28.404885

In [303]:
# Plot a line chart with all the inertia values computed with the different values of k to visually identify the optimal value for k.
elbow_plot = elbow_data_df.hvplot.line(
    x='k',
    y='inertia',
    title='Elbow Method for Optimal k',
    xlabel='Number of Clusters (k)',
    ylabel='Inertia',
    width=800,
    height=400
)

# Display the plot
elbow_plot


:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 3

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [304]:
#Set the optimal number of clusters
optimal_k = 3  

# Initialize the K-Means model using the best value for k
kmeans_model = KMeans(n_clusters=optimal_k, random_state=42) 


In [305]:
# Fit the K-Means model using the scaled DataFrame
kmeans_model.fit(df_scaled)

C:\Users\danie\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=3, random_state=42)

In [306]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
clusters = kmeans_model.predict(df_scaled)


In [307]:
# Create a copy of the scaled DataFrame



In [327]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
df_scaled['Cluster'] = clusters #Create a new DataFrame to hold the scaled data along with the predicted clusters

# include the original coin_id to reference back to the original data
df_scaled['coin_id'] = df_market_data.index

# Display the copy of the scaled DataFrame
df_scaled.head()


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  Cluster       coin_id  
coin_id                                                          
bitcoin                        -0.251637        0       bitcoin  
ethereum                       -0.199352        0      ethereum  
tether                         -0.282061        2        tether  
ripple                         -0.295546        2        ripple  
bitcoin-cash                   -0.270317        0  bitcoin-cash

In [328]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.

# Create a scatter plot using hvPlot
scatter_plot = df_scaled.hvplot.scatter(
    x="price_change_percentage_24h",  # X-axis
    y="price_change_percentage_7d",    # Y-axis
    by="Cluster",                       # Color by the cluster labels
    hover_cols=["coin_id"],            # Show the cryptocurrency name on hover
    title="Cryptocurrency Clusters by Price Change",
    width=800,
    height=400
)

# Display the scatter plot
scatter_plot



:NdOverlay   [Cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [310]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)  # Reduce to 3 components for visualization


In [330]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame down to three principal components.
pca_result = pca.fit_transform(df_scaled[numerical_cols])  # Use only numerical columns for PCA

# Create a DataFrame for PCA results
df_pca = pd.DataFrame(pca_result, columns=['PCA1', 'PCA2', 'PCA3'], index=df_scaled.index)  # Update to include PCA3

# Add the cluster labels to the PCA DataFrame
df_pca.head()


PCA1      PCA2      PCA3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

In [331]:
# Retrieve the explained variance to determine how much information can be attributed to each principal component.
explained_variance = pca.explained_variance_ratio_

# Create a DataFrame to display the explained variance
explained_variance_df = pd.DataFrame({
    'Principal Component': [f'PC{i+1}' for i in range(len(explained_variance))],
    'Explained Variance Ratio': explained_variance
})

# Display the explained variance DataFrame
explained_variance_df


Principal Component  Explained Variance Ratio
0                 PC1                  0.371986
1                 PC2                  0.347008
2                 PC3                  0.176038

In [332]:
# Total explained variance for the selected principal components
total_explained_variance = explained_variance.sum()

# Display the total explained variance
total_explained_variance

0.8950316570309841

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.895

In [334]:
# Create a new DataFrame with the PCA data.
df_pca = pd.DataFrame(pca_result, columns=['PCA1', 'PCA2', 'PCA3'], index=df_scaled.index)

# Copy the crypto names from the original scaled DataFrame
df_pca['Cluster'] = df_scaled['Cluster']
df_pca['coin_id'] = df_scaled.index

# Set the coin_id column as the index
df_pca.set_index('coin_id', inplace=True)

# Display the scaled PCA DataFrame
df_pca.head()

PCA1      PCA2      PCA3  Cluster
coin_id                                            
bitcoin      -0.600667  0.842760  0.461595        0
ethereum     -0.458261  0.458466  0.952877        0
tether       -0.433070 -0.168126 -0.641752        2
ripple       -0.471835 -0.222660 -0.479053        2
bitcoin-cash -1.157800  2.041209  1.859715        0

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [335]:
# Create a list with the number of k-values from 1 to 11
k_values = range(1, 11)

In [337]:
# Create an empty list to store the inertia values
inertia = [] #Create a list to store inertia values

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list

#Loop through the range of k values
for k in k_values:
    # 1. Create a KMeans model using the loop counter for the n_clusters
    model = KMeans(n_clusters=k, random_state=42)
    
    # Fit the model to the scaled PCA data
    model.fit(df_pca[['PCA1', 'PCA2', 'PCA3']])  #2. Fit the model to the data using `df_market_data_pca`
    
    # Append the model.inertia_ to the inertia list
    inertia.append(model.inertia_)

C:\Users\danie\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\danie\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\danie\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\danie\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is k

In [338]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_dict = dict(zip(k_values, inertia))

# Create a DataFrame with the data to plot the Elbow curve
elbow_data_df = pd.DataFrame(list(elbow_data_dict.items()), columns=['k', 'inertia'])
elbow_data_df

k     inertia
0   1  256.874086
1   2  182.339530
2   3  140.431833
3   4   49.665497
4   5   42.051901
5   6   31.749005
6   7   28.031869
7   8   19.451747
8   9   13.814082
9  10   10.630648

In [339]:
# Plot a line chart with all the inertia values computed with the different values of k to visually identify the optimal value for k.
elbow_plot = elbow_data_df.hvplot.line(
    x='k',
    y='inertia',
    title='Elbow Method for Optimal k',
    xlabel='Number of Clusters (k)',
    ylabel='Inertia',
    width=800,
    height=400
)

# Display the plot
elbow_plot

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** Yes, the new calculation suggests that data points are closer to their respective cluster centers

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [350]:
# Set the optimal number of clusters
optimal_k = 4  

#Initialize the K-Means model using the best value for k
kmeans_model_pca = KMeans(n_clusters=optimal_k, random_state=42)  

In [351]:
# Fit the K-Means model using the PCA data
kmeans_model_pca.fit(df_pca[['PCA1', 'PCA2', 'PCA3']])  

C:\Users\danie\anaconda3\envs\dev\lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=42)

In [352]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
clusters_pca = kmeans_model_pca.predict(df_pca[['PCA1', 'PCA2', 'PCA3']])

# Print the resulting array of cluster values.
clusters_pca

array([0, 0, 2, 2, 0, 0, 0, 0, 0, 2, 2, 2, 2, 0, 2, 0, 2, 2, 0, 2, 2, 0,
       2, 2, 2, 2, 2, 2, 0, 2, 2, 2, 3, 0, 2, 2, 1, 2, 2, 2, 2])

In [353]:
# Create a copy of the scaled PCA DataFrame

# Create a new DataFrame to hold the PCA data along with the predicted clusters
# Add a new column to the copy of the PCA DataFrame with the predicted clusters
df_pca['Cluster'] = clusters_pca

# Include the original coin_id to reference back to the original data
df_pca['coin_id'] = df_market_data.index


# Display the copy of the scaled PCA DataFrame
df_pca.head()

PCA1      PCA2      PCA3  Cluster       coin_id
coin_id                                                          
bitcoin      -0.600667  0.842760  0.461595        0       bitcoin
ethereum     -0.458261  0.458466  0.952877        0      ethereum
tether       -0.433070 -0.168126 -0.641752        2        tether
ripple       -0.471835 -0.222660 -0.479053        2        ripple
bitcoin-cash -1.157800  2.041209  1.859715        0  bitcoin-cash

In [354]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.


scatter_plot_pca = df_pca.hvplot.scatter(
    x="PCA1",  
    y="PCA2",  
    by="Cluster",  # Color by the cluster labels
    hover_cols=["coin_id"],  # Show the cryptocurrency name on hover
    title="Cryptocurrency Clusters in PCA Space",
    width=800,
    height=400
)

# Display the scatter plot
scatter_plot_pca

:NdOverlay   [Cluster]
   :Scatter   [PCA1]   (PCA2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [355]:
# Composite plot to contrast the Elbow curves
# YOUR CODE HERE!

# Combine the two DataFrames for plotting
elbow_data_df_scaled['Method'] = 'Scaled Data'
elbow_data_df_pca['Method'] = 'PCA Data'

# Concatenate both DataFrames
elbow_data_df_combined = pd.concat([elbow_data_df_scaled, elbow_data_df_pca], ignore_index=True)

# Plot the composite elbow curves
elbow_plot = elbow_data_df_combined.hvplot.line(
    x='k',
    y='inertia',
    by='Method',
    title='Elbow Method Comparison: Scaled Data vs PCA Data',
    xlabel='Number of Clusters (k)',
    ylabel='Inertia',
    width=800,
    height=400
)

# Display the plot
elbow_plot

:NdOverlay   [Method]
   :Curve   [k]   (inertia)

In [357]:
# Composite plot to contrast the clusters
# YOUR CODE HERE!

# Combine the plots using hvPlot's layout feature
composite_plot = (scatter_plot + scatter_plot_pca).cols(2)  

# Display the composite plot
composite_plot

:Layout
   .NdOverlay.I  :NdOverlay   [Cluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [Cluster]
      :Scatter   [PCA1]   (PCA2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** The data looks more dispersed with a K=3 rather than a k=4. However, a k=4 shows that some data is treated as outliers.